Support for disabling PLT for better function call performance

GabrielMajeri opened the pull request:
This PR gives `rustc` the ability to skip the PLT when generating function calls into shared libraries. This can improve performance by reducing branch indirection.

AFAIK, the only advantage of using the PLT is to allow for ELF lazy binding. However, since Rust already [enables full relro for security](https://github.com/rust-lang/rust/pull/43170), lazy binding was disabled anyway.

This is a little known feature which is supported by [GCC](https://gcc.gnu.org/onlinedocs/gcc/Code-Gen-Options.html) and [Clang](https://clang.llvm.org/docs/ClangCommandLineReference.html#cmdoption-clang-fplt) as `-fno-plt` (some Linux distros [enable it by default](https://git.archlinux.org/svntogit/packages.git/tree/trunk/makepkg.conf?h=packages/pacman#n40) for all builds).

Implementation inspired by [this patch](https://reviews.llvm.org/D39079#change-YvkpNDlMs_LT) which adds `-fno-plt` support to Clang.

## Performance

I didn't run a lot of benchmarks, but these are the results on my machine for a `clap` [benchmark](https://github.com/clap-rs/clap/blob/master/benches/05_ripgrep.rs):

```
 name              control ns/iter  no-plt ns/iter  diff ns/iter  diff %  speedup 
 build_app_long    11,097           10,733                  -364  -3.28%   x 1.03 
 build_app_short   11,089           10,742                  -347  -3.13%   x 1.03 
 build_help_long   186,835          182,713               -4,122  -2.21%   x 1.02 
 build_help_short  80,949           78,455                -2,494  -3.08%   x 1.03 
 parse_clean       12,385           12,044                  -341  -2.75%   x 1.03 
 parse_complex     19,438           19,017                  -421  -2.17%   x 1.02 
 parse_lots        431,493          421,421              -10,072  -2.33%   x 1.02 
```

A small performance improvement across the board, with no downsides. It's likely binaries which make a lot of function calls into dynamic libraries could see even more improvements. [This comment](https://patchwork.ozlabs.org/patch/468993/#1028255) suggests that, in some cases, `-fno-plt` could improve PIC/PIE code performance by 10%.

## Security benefits

**Bonus**: some of the speculative execution attacks rely on the PLT, by disabling it we reduce a big attack surface and reduce the need for [`retpoline`](https://reviews.llvm.org/D41723).

## Remaining PLT calls

The compiled binaries still have plenty of PLT calls, coming from C/C++ libraries. Building dependencies with `CFLAGS=-fno-plt CXXFLAGS=-fno-plt` removes them.

GabrielMajeri committed:
Support for disabling the PLT on ELF targets

Disable the PLT where possible to improve performance
for indirect calls into shared libraries.

This optimization is enabled by default where possible.

- Add the `NonLazyBind` attribute to `rustllvm`:
  This attribute informs LLVM to skip PLT calls in codegen.

- Disable PLT unconditionally:
  Apply the `NonLazyBind` attribute on every function.

- Only enable no-plt when full relro is enabled:
  Ensures we only enable it when we have linker support.

- Add `-Z plt` as a compiler option

rust-highfive commented:
r? @nikomatsakis

(rust_highfive has picked a reviewer for you, use r? to override)

eddyb commented:
cc @alexcrichton

cuviper commented:
> AFAIK, the only advantage of using the PLT is to allow for ELF lazy binding. However, since Rust already [enables full relro for security](https://github.com/rust-lang/rust/pull/43170), lazy binding was disabled anyway.

Note that PPC64 is only defaulting to partial relro due to an old ld.so bug in bind-now.
https://github.com/rust-lang/rust/pull/43170/files#diff-b2d51315427bd679ca33d47167e82171R20

There's also an option for `-Z relro-level={full,partial,off}`.  I'll try to see if I see similar PPC64 issues arise with `-fno-plt`, but my initial feeling is that we should only enable this in conjunction with `relro-level=full`.

rust-highfive commented:
The job `x86_64-gnu-llvm-5.0` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/433626813) ([raw log](https://api.travis-ci.org/v3/job/433626813/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
[00:55:36] ....................................................................................................
[00:55:39] ..............................................................i.....................................
[00:55:42] ....................................................................................................
[00:55:45] ....................................................................................................
[00:55:48] ...........iiiiiiiii................................................................................
[00:55:53] ....................................................................................................
[00:55:57] ...............................................................................................i....
[00:56:00] ....................................................................................................
[00:56:03] .......................................................i.i..ii......................................
---
travis_time:start:test_codegen
Check compiletest suite=codegen mode=codegen (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
[01:04:25] 
[01:04:25] running 107 tests
[01:04:28] i..ii...iii....i...i............iii...........i....Fi....ii...i.i.ii..............i...ii..ii.i....ii
[01:04:28] thread 'main' panicked at 'Some tests failed', tools/compiletest/src/main.rs:496:22
[01:04:28] failures:
[01:04:28] 
[01:04:28] ---- [codegen] codegen/naked-functions.rs stdout ----
[01:04:28] 
[01:04:28] 
[01:04:28] error: verification with 'FileCheck' failed
[01:04:28] status: exit code: 1
[01:04:28] command: "/usr/lib/llvm-5.0/bin/FileCheck" "--input-file" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/naked-functions/naked-functions.ll" "/checkout/src/test/codegen/naked-functions.rs"
[01:04:28] ------------------------------------------
[01:04:28] 
[01:04:28] ------------------------------------------
[01:04:28] stderr:
[01:04:28] stderr:
[01:04:28] ------------------------------------------
[01:04:28] /checkout/src/test/codegen/naked-functions.rs:18:11: error: expected string not found in input
[01:04:28] // CHECK: Function Attrs: naked uwtable
[01:04:28]           ^
[01:04:28] /checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/naked-functions/naked-functions.ll:1:1: note: scanning from here
[01:04:28] ; ModuleID = 'naked_functions.3a1fbbbh-cgu.0'
[01:04:28] ^
[01:04:28] /checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/naked-functions/naked-functions.ll:6:3: note: possible intended match here
[01:04:28] ; Function Attrs: naked nonlazybind uwtable
[01:04:28] 
[01:04:28] ------------------------------------------
[01:04:28] 
[01:04:28] thread '[codegen] codegen/naked-functions.rs' panicked at 'explicit panic', tools/compiletest/src/runtest.rs:3238:9
---
[01:04:28] test result: FAILED. 77 passed; 1 failed; 29 ignored; 0 measured; 0 filtered out
[01:04:28] 
[01:04:28] 
[01:04:28] 
[01:04:28] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--src-base" "/checkout/src/test/codegen" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--mode" "codegen" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-5.0/bin/FileCheck" "--host-rustcflags" "-Crpath -O -Zunstable-options " "--target-rustcflags" "-Crpath -O -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python2.7" "--lldb-python" "/usr/bin/python2.7" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "5.0.0\n" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--llvm-components" "" "--llvm-cxxflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--color" "always"
[01:04:28] 
[01:04:28] 
[01:04:28] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test
[01:04:28] Build completed unsuccessfully in 0:17:44
[01:04:28] Build completed unsuccessfully in 0:17:44
[01:04:28] Makefile:58: recipe for target 'check' failed
[01:04:28] make: *** [check] Error 1

The command "stamp sh -x -c "$RUN_SCRIPT"" exited with 2.
travis_time:start:0d3ba4a0
$ date && (curl -fs --head https://google.com | grep ^Date: | sed 's/Date: //g' || true)
---
travis_time:end:0204f06f:start=1537985051202083688,finish=1537985051358059111,duration=155975423
travis_fold:end:after_failure.4
travis_fold:start:after_failure.5
travis_time:start:1243b65a
$ cat ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers || true
cat: ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers: No such file or directory
travis_fold:end:after_failure.5
travis_fold:start:after_failure.6
travis_time:start:23d227fe
$ dmesg | grep -i kill
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

nagisa commented:
Please add a flag that controls this behaviour. For now it can be a debug `-Z` flag, similar to other such flags (e.g. `-Zmutable-noalias`).

> Also, I'm not sure how to fix the failing test. Is it possible for CHECK: Function Attrs to allow extra attributes in the output, besides the ones being tested?

CHECK lines only check for matching line prefix. That test in particular seems to be testing for `naked` only, in which case you can probably remove the other attribute from the `CHECK` line to have it pass. Alternatively, you might have some success with [pattern matching syntax](https://llvm.org/docs/CommandGuide/FileCheck.html#filecheck-pattern-matching-syntax).

nagisa commented:
> objdump says that the number of function calls using the PLT goes down, but a lot of functions calls are still using the PLT. So I'm guessing I have to add this in other places too, but I'm not very familiar with the code.

A large number of `@PLT` symbols likely come from outside the rust ecosystem (e.g. glibc, llvm, etc.). Those might need to be taken care of independently (by changing build system configuration, perhaps?). You might want to submit a similar patch to the [cc crate](https://github.com/alexcrichton/cc-rs).

---

(Addressed to not-author, but somebody who knows how to do perf runs) I also think a perf run would be great, but not sure how to start it.

varkor commented:
@bors try

bors commented:
:hourglass: Trying commit ddf98c1f387301842714a5a2d92267ea20c49669 with merge 57476313f10d8c8068ce4241d959e228fc7eb6a3...

rust-highfive commented:
The job `x86_64-gnu-llvm-5.0` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/433672174) ([raw log](https://api.travis-ci.org/v3/job/433672174/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
[00:58:24] ....................................................................................................
[00:58:27] ..............................................................i.....................................
[00:58:30] ....................................................................................................
[00:58:33] ....................................................................................................
[00:58:36] ............iiiiiiiii...............................................................................
[00:58:42] ....................................................................................................
[00:58:46] ...............................................................................................i....
[00:58:49] ....................................................................................................
[00:58:52] .......................................................i.i..ii......................................
---
travis_time:start:test_codegen
Check compiletest suite=codegen mode=codegen (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
[01:07:32] 
[01:07:32] running 107 tests
[01:07:35] i..ii...iii....i...i............iii...........i.....iF...ii...i.i.ii..............i...ii..ii.i....ii
[01:07:35] thread 'main' panicked at 'Some tests failed', tools/compiletest/src/main.rs:496:22
[01:07:35] failures:
[01:07:35] 
[01:07:35] ---- [codegen] codegen/naked-functions.rs stdout ----
[01:07:35] 
[01:07:35] 
[01:07:35] error: verification with 'FileCheck' failed
[01:07:35] status: exit code: 1
[01:07:35] command: "/usr/lib/llvm-5.0/bin/FileCheck" "--input-file" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/naked-functions/naked-functions.ll" "/checkout/src/test/codegen/naked-functions.rs"
[01:07:35] ------------------------------------------
[01:07:35] 
[01:07:35] ------------------------------------------
[01:07:35] stderr:
[01:07:35] stderr:
[01:07:35] ------------------------------------------
[01:07:35] /checkout/src/test/codegen/naked-functions.rs:18:11: error: expected string not found in input
[01:07:35] // CHECK: Function Attrs: naked uwtable
[01:07:35]           ^
[01:07:35] /checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/naked-functions/naked-functions.ll:1:1: note: scanning from here
[01:07:35] ; ModuleID = 'naked_functions.3a1fbbbh-cgu.0'
[01:07:35] ^
[01:07:35] /checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/naked-functions/naked-functions.ll:6:3: note: possible intended match here
[01:07:35] ; Function Attrs: naked nonlazybind uwtable
[01:07:35] 
[01:07:35] ------------------------------------------
[01:07:35] 
[01:07:35] thread '[codegen] codegen/naked-functions.rs' panicked at 'explicit panic', tools/compiletest/src/runtest.rs:3238:9
---
[01:07:35] test result: FAILED. 77 passed; 1 failed; 29 ignored; 0 measured; 0 filtered out
[01:07:35] 
[01:07:35] 
[01:07:35] 
[01:07:35] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--src-base" "/checkout/src/test/codegen" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--mode" "codegen" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-5.0/bin/FileCheck" "--host-rustcflags" "-Crpath -O -Zunstable-options " "--target-rustcflags" "-Crpath -O -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python2.7" "--lldb-python" "/usr/bin/python2.7" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "5.0.0\n" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--llvm-components" "" "--llvm-cxxflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--color" "always"
[01:07:35] 
[01:07:35] 
[01:07:35] failed to run: /checkout/obj/build/756 ./src/tools/lldb/www
37080 ./obj/build/x86_64-unknown-linux-gnu/stage0-std/release
---
travis_time:end:0c6f72ba:start=1537990447136677365,finish=1537990447141035866,duration=4358501
travis_fold:end:after_failure.3
travis_fold:start:after_failure.4
travis_time:start:017849be
$ ln -s . checkout && for CORE in obj/cores/core.*; do EXE=$(echo $CORE | sed 's|obj/cores/core\.[0-9]*\.!checkout!\(.*\)|\1|;y|!|/|'); if [ -f "$EXE" ]; then print
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

varkor commented:
@rust-timer build https://github.com/rust-lang/rust/commit/57476313f10d8c8068ce4241d959e228fc7eb6a3

rust-timer commented:
Please provide the full 40 character commit hash.

GabrielMajeri commented:
Perf results are in, nice improvements on wall time. From what I've seen, the patch currently only removes *about 20%* of the total PLT calls, there's probably still some more performance to be gained. 

@nagisa 
> Please add a flag that controls this behaviour. 

Alright, I've implemented a `-Z no-plt` flag, but should this option be enabled or disabled by default?

Also, what should we do if the user specified `-Z no-plt`, but the flag is then ignored, because full relro isn't supported (for example, as cuviper mentioned, linker issues on PowerPC64)?

> symbols likely come from outside the rust ecosystem

Thanks for the tip, but I'm still unable to get rid of the PLT. I've added `CFLAGS=-fno-plt` to my system, and rebuilt the compiler from source, but `rustc` still generates _lots_ of calls which use the PLT.

If I build C binaries on my system, the final binary doesn't even have a `.plt` section, it is completly removed.

cuviper commented:
> Alright, I've implemented a `-Z no-plt` flag, but should this option be enabled or disabled by default?

I think the way you've documented it is fine, "(default: PLT is disabled if full relro is enabled)".

> Also, what should we do if the user specified `-Z no-plt`, but the flag is then ignored, because full relro isn't supported (for example, as cuviper mentioned, linker issues on PowerPC64)?

Don't ignore it.  These are advanced options -- if the user asks for `plt=off` without full relro, let them deal with the implications.  So the check is something like `plt.unwrap_or(relro != Full)`.

cuviper commented on src/librustc/session/config.rs:
```diff
@@ -1387,6 +1387,10 @@ options! {DebuggingOptions, DebuggingSetter, basic_debugging_options,
           "output a json file with profiler results"),
     emit_stack_sizes: bool = (false, parse_bool, [UNTRACKED],
           "emits a section containing stack size metadata"),
+    plt: Option<bool> = (Some(false), parse_opt_bool, [TRACKED],
```
Shouldn't this be initialized `None`, so we know whether it has been explicitly set?

nagisa replied:
Yes, this should be a `None`. While `Some(false)` makes a long of sense with the current defaults, it is easier to make changes to defaults and code overall, when there are no surprises between “not provided” and “`-Zplt=no` is provided”.

nikomatsakis commented:
r? @nagisa -- since they had some concerns :)

joshtriplett commented:
This looks great to me; it's a good optimization with no downsides.

cuviper commented:
BTW, one case where you do need PLT is to use tools like [ltrace](http://man7.org/linux/man-pages/man1/ltrace.1.html) and [latrace](http://people.redhat.com/jolsa/latrace/latrace.html) which hook PLT calls.  But I think they're uncommon enough that it's OK for us to take the performance win by default, and have the `-Z plt` option when you really want to use those tools.

nagisa commented:
Is it possible to warn if the options are conflicting e.g. `-Zplt=yes` and relro is not full? It would be confusing to have `-Zplt=yes` silently ignored.

nagisa commented:
Or rather, I should be asking what is the failure mode when `-Zplt=yes` and “relro” is not “full”? I see that the code only takes relro into account only when PLT usage is not forced one way or the other.

nagisa commented:
Also, sorry to be making so many comments at once, but if you ping me on IRC/Zulip (same nick as here on both), we could work together on figuring out where the rest of the PLT calls come from.

nagisa commented:
Doing another perf run – there is now way fewer `@PLT` calls since the last run.

@bors try

bors commented:
:hourglass: Trying commit 38ced017bc41b12bf82e4aea96eb52cd81bcc0e3 with merge 0118a01c666afd6313eda8a0cb34e676b8e33c4c...

bors commented:
:broken_heart: Test failed - [status-travis](https://travis-ci.org/rust-lang/rust/builds/434744800?utm_source=github_status&utm_medium=notification)

rust-highfive commented:
The job `dist-x86_64-linux` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/434744801) ([raw log](https://api.travis-ci.org/v3/job/434744801/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
travis_fold:end:services

travis_fold:start:git.checkout
travis_time:start:06001b4c
$ git clone --depth=2 --branch=try https://github.com/rust-lang/rust.git rust-lang/rust
---
[00:36:22] -- Installing: /checkout/obj/build/x86_64-unknown-linux-gnu/llvm/include/llvm/ADT/Optional.h
[00:36:22] -- Installing: /checkout/obj/build/x86_64-unknown-linux-gnu/llvm/include/llvm/ADT/SmallString.h
[00:36:22] -- Installing: /checkout/obj/build/x86_64-unknown-linux-gnu/llvm/include/llvm/ADT/SparseMultiSet.h
[00:36:22] -- Installing: /checkout/obj/build/x86_64-unknown-linux-gnu/llvm/include/llvm/ADT/ImmutableMap.h
/home/travis/.travis/job_stages: line 348: echo: write error: Resource temporarily unavailable

The command "date && (curl -fs --head https://google.com | grep ^Date: | sed 's/Date: //g' || true)
" exited with 1.
travis_fold:start:after_failure.1
---
travis_time:end:24fcc040:start=1538161768042357676,finish=1538161768048374750,duration=6017074
travis_fold:end:after_failure.3
travis_fold:start:after_failure.4
travis_time:start:1f94c361
$ ln -s . checkout && for CORE in obj/cores/core.*; do EXE=$(echo $CORE | sed 's|obj/cores/core\.[0-9]*\.!checkout!\(.*\)|\1|;y|!|/|'); if [ -f "$EXE" ]; then printf travis_fold":start:crashlog\n\033[31;1m%s\033[0m\n" "$CORE"; gdb --batch -q -c "$CORE" "$EXE" -iex 'set auto-load off' -iex 'dir src/' -iex 'set sysroot .' -ex bt -ex q; echo travis_fold":"end:crashlog; fi; done || true
travis_fold:end:after_failure.4
travis_fold:start:after_failure.5
travis_time:start:0dd37f8e
travis_time:start:0dd37f8e
$ cat ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers || true
cat: ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers: No such file or directory
travis_fold:end:after_failure.5
travis_fold:start:after_failure.6
travis_time:start:29aaa443
$ dmesg | grep -i kill
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

alexcrichton commented:
@bors: retry

bors commented:
:hourglass: Trying commit 38ced017bc41b12bf82e4aea96eb52cd81bcc0e3 with merge 495deb00b88edd27db21ba2366c0e85da25db407...

bors commented:
:broken_heart: Test failed - [status-travis](https://travis-ci.org/rust-lang/rust/builds/434788902?utm_source=github_status&utm_medium=notification)

alexcrichton commented:
Hm this'll probably need a force push to re-try unfortunately

GabrielMajeri commented:
@alexcrichton I've rebased and force-updated.

nagisa commented:
@bors try

bors commented:
:hourglass: Trying commit 76c4031af71b381d7c74914815d789a96322ae50 with merge f4a266585ffc702dfdbfafdec11e09f6128d6c38...

bors commented:
:sunny: Test successful - [status-travis](https://travis-ci.org/rust-lang/rust/builds/434921726?utm_source=github_status&utm_medium=notification)
State: approved= try=True

nagisa commented:
@rust-timer build f4a266585ffc702dfdbfafdec11e09f6128d6c38

rust-timer commented:
Success: Queued f4a266585ffc702dfdbfafdec11e09f6128d6c38 with parent 6622172734a6c456ea85e17b765ad1dbc0e42907, [comparison URL](https://perf.rust-lang.org/compare.html?start=6622172734a6c456ea85e17b765ad1dbc0e42907&end=f4a266585ffc702dfdbfafdec11e09f6128d6c38).

Valinora commented:
Perf is in, results seem mixed but mostly positive

nagisa commented:
r=me on the implementation, just waiting for the checkboxes.

nagisa commented:
@bors r+

bors commented:
:pushpin: Commit 76c4031af71b381d7c74914815d789a96322ae50 has been approved by `nagisa`

<!-- @bors r=nagisa 76c4031af71b381d7c74914815d789a96322ae50 -->

nagisa commented:
@GabrielMajeri if you’re around it would be great if you could squash commits a little. “Fix failing test” or “rename flag” have very little value as separate commits.

nagisa commented:
@bors r+

bors commented:
:pushpin: Commit 7a190e060bd949b976e67826962b12028b0438d1 has been approved by `nagisa`

<!-- @bors r=nagisa 7a190e060bd949b976e67826962b12028b0438d1 -->

bors commented:
:hourglass: Testing commit 7a190e060bd949b976e67826962b12028b0438d1 with merge 7287f9da0043fa2bb28e1d20a039a6dcc4295ad9...

bors commented:
:broken_heart: Test failed - [status-travis](https://travis-ci.org/rust-lang/rust/builds/439754795?utm_source=github_status&utm_medium=notification)

rust-highfive commented:
The job `dist-various-2` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/439754807) ([raw log](https://api.travis-ci.org/v3/job/439754807/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
[00:53:50]    Compiling unwind v0.0.0 (/checkout/src/libunwind)
[00:53:50]    Compiling compiler_builtins v0.0.0 (/checkout/src/rustc/compiler_builtins_shim)
[00:53:50]    Compiling alloc_jemalloc v0.0.0 (/checkout/src/liballoc_jemalloc)
[00:54:14]    Compiling std v0.0.0 (/checkout/src/libstd)
[00:54:21] LLVM ERROR: Cannot select: 0x7fe8e97384e0: ch,glue = X86ISD::CALL 0x7fe8e9502958, 0x7fe8e95022d8, Register:i32 $edi, RegisterMask:Untyped, 0x7fe8e9502958:1, libcore/ptr.rs:1563:12
[00:54:21]   0x7fe8e95022d8: i32,ch = load<(load 4 from got)> 0x7fe8e9614758, 0x7fe8e9502680, undef:i32, libcore/ptr.rs:1563:12
[00:54:21]     0x7fe8e9502680: i32 = X86ISD::WrapperRIP TargetGlobalAddress:i32<void ({ [0 x i32], { [0 x i8]*, i32 }, [0 x i32], { [0 x i8]*, i32 }, [0 x i32], i32, [0 x i32], i32, [0 x i32] }*)* @_ZN4core9panicking5panic17he0a0446e9b6bf934E> 0 [TF=5], libcore/ptr.rs:1563:12
[00:54:21]       0x7fe8e9502138: i32 = TargetGlobalAddress<void ({ [0 x i32], { [0 x i8]*, i32 }, [0 x i32], { [0 x i8]*, i32 }, [0 x i32], i32, [0 x i32], i32, [0 x i32] }*)* @_ZN4core9panicking5panic17he0a0446e9b6bf934E> 0 [TF=5], libcore/ptr.rs:1563:12
[00:54:21]     0x7fe8e95020d0: i32 = undef
[00:54:21]   0x7fe8e95027b8: i32 = Register $edi
[00:54:21]   0x7fe8e95021a0: Untyped = RegisterMask
[00:54:21]   0x7fe8e9502958: ch,glue = CopyToReg 0x7fe8e95025b0, Register:i32 $edi, 0x7fe8e95026e8, libcore/ptr.rs:1563:12
[00:54:21]     0x7fe8e95027b8: i32 = Register $edi
[00:54:21]     0x7fe8e95026e8: i32 = X86ISD::WrapperRIP TargetGlobalAddress:i32<<{ i8*, [4 x i8], i8*, [12 x i8] }>* @anon.edb5cb93134b159a66f2de0e3c5659b0.2.llvm.10114106681850442039> 0, libcore/ptr.rs:1563:12
[00:54:21]       0x7fe8e9502af8: i32 = TargetGlobalAddress<<{ i8*, [4 x i8], i8*, [12 x i8] }>* @anon.edb5cb93134b159a66f2de0e3c5659b0.2.llvm.10114106681850442039> 0, libcore/ptr.rs:1563:12
[00:54:21] In function: _ZN4core3ptr33_$LT$impl$u20$$BP$const$u20$T$GT$12align_offset17h88359f35680ecfe2E
[00:54:21] error: Could not compile `core`.
[00:54:21] 
[00:54:21] To learn more, run the command again with --verbose.
[00:54:21] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "build" "--target" "x86_64-unknown-linux-gnux32" "-j" "4" "--release" "--locked" "--color" "always" "--features" "panic-unwind jemalloc backtrace" "--manifest-path" "/checkout/src/libstd/Cargo.toml" "--message-format" "json"
[00:54:21] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "build" "--target" "x86_64-unknown-linux-gnux32" "-j" "4" "--release" "--locked" "--color" "always" "--features" "panic-unwind jemalloc backtrace" "--manifest-path" "/checkout/src/libstd/Cargo.toml" "--message-format" "json"
[00:54:21] expected success, got: exit code: 101
[00:54:21] thread 'main' panicked at 'cargo must succeed', bootstrap/compile.rs:1112:9
[00:54:21] travis_fold:end:stage2-std

[00:54:21] travis_time:end:stage2-std:start=1539196004625854356,finish=1539196036058284933,duration=31432430577

---
travis_time:end:0bda4d28:start=1539196037257181001,finish=1539196037264406946,duration=7225945
travis_fold:end:after_failure.3
travis_fold:start:after_failure.4
travis_time:start:01ab913a
$ ln -s . checkout && for CORE in obj/cores/core.*; do EXE=$(echo $CORE | sed 's|obj/cores/core\.[0-9]*\.!checkout!\(.*\)|\1|;y|!|/|'); if [ -f "$EXE" ]; then printf travis_fold":start:crashlog\n\033[31;1m%s\033[0m\n" "$CORE"; gdb --batch -q -c "$CORE" "$EXE" -iex 'set auto-load off' -iex 'dir src/' -iex 'set sysroot .' -ex bt -ex q; echo travis_fold":"end:crashlog; fi; done || true
travis_fold:end:after_failure.4
travis_fold:start:after_failure.5
travis_time:start:065694e7
travis_time:start:065694e7
$ cat ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers || true
cat: ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers: No such file or directory
travis_fold:end:after_failure.5
travis_fold:start:after_failure.6
travis_time:start:115ee313
$ dmesg | grep -i kill
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

GabrielMajeri commented:
Well I don't know what's wrong with `gnux32`, but in my defense [Clang  also fails to compile](https://gcc.godbolt.org/z/ALEsEJ) when I pass `-fPIC -fno-plt` on the `gnux32` target.
And it's [apparently a know LLVM bug](https://bugs.llvm.org/show_bug.cgi?id=36743).

Should we just blacklist gnux32 for the  `-Z plt` option?

nagisa commented:
Simply ignoring the test for the target should be fine.

cuviper commented:
> Simply ignoring the test for the target should be fine.

Wouldn't that still mean the _default_ no-plt is broken on gnux32?

nagisa commented:
Oh right, I neglected to remember we’re making `-Zplt=off` by default.

In that case the default should be indicated in the target specification and the `gnux32` should explicitly enable plt by default.

Thinking about it now, indicating PLT default in the target specification is a good idea nevertheless, but unlike with the flag target specifications aren’t covered by the stability, and would become insta-stable…

GabrielMajeri commented:
The way I see it, `-Z plt=off` is an optimization, we don't guarantee it does anything (it's a best effort kind of thing). For now, I changed the code to unconditionally disable the optimization on `gnux32`, at least until LLVM gets fixed.

nagisa commented:
This should still be a part of target specification because there are
targwta defined outside of the compiler and we do t test them all.

On Thu, Oct 11, 2018, 06:57 Gabriel Majeri <notifications@github.com> wrote:

> The way I see it, -Z plt=off is an optimization, we don't guarantee it
> does anything (it's a best effort kind of thing). For now, I changed the
> code to unconditionally disable the optimization on gnux32, at least
> until LLVM gets fixed.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rust-lang/rust/pull/54592#issuecomment-428812128>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AApc0qa2Eaq3zyK0cBG93zjHz_aFFJelks5ujsGYgaJpZM4W7FzR>
> .
>

GabrielMajeri commented:
@nagisa As far as I understand, even if somebody defines some external target with this ABI, there shouldn't be an issue.

The code checks for the (custom) target's `llvm-target` attribute to see if it contains `gnux32`. This is as far as we know the only ABI where LLVM currently has an issue (due to a bug).

For example, Clang accepts this option for all targets and ABIs (even Windows). On targets where it doesn't do anything, it emits the attributes, and LLVM just ignores them (except for this buggy ABI which crashes).

nagisa commented:
I'm arguing that the author of external target specifications should have a
full control over such matters.

While gnux32 is the only target we know that crashes, our triplet coverage
is incomplete and we can't say for sure that any future triplets llvm may
support will be problem free either.

In the past platform specific settings and workarounds were always stored
within the target files and I see no reason whatsoever for us to change
that now. And it is strictly more flexible than checking whether llvm
triple string ends in a specific substring.

I also don't see much of an issue with allowing people to force arbitrary
configurations (i.e. allowing them to specify -Zplt=off) even if we know
they might not be working, ignored, irrelevant etc. At worst they'll
encounter an ICE. In fact, if they do explicitly request for some
behaviour, I believe that should override any workaround or default (which
is not what happens with the current iteration of the PR)







On Thu, Oct 11, 2018, 08:39 Gabriel Majeri <notifications@github.com> wrote:

> @nagisa <https://github.com/nagisa> As far as I understand, even if
> somebody defines some external target with this ABI, there shouldn't be an
> issue.
>
> The code checks for the (custom) target's llvm-target attribute to see if
> it contains gnux32. This is as far as we know the only ABI where LLVM
> currently has an issue (due to a bug).
>
> For example, Clang accepts this option for all targets and ABIs (even
> Windows). On targets where it doesn't do anything, it emits the attributes,
> and LLVM just ignores them (except for this buggy ABI which crashes).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rust-lang/rust/pull/54592#issuecomment-428827134>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AApc0mkkFmjCAO_7YT5ecCUdrWQai3liks5ujtmQgaJpZM4W7FzR>
> .
>

pnkfelix commented:
(@nagisa said at T-compiler meeting that we can un-nominate this)

GabrielMajeri commented:
@nagisa ok, I've added a `needs_plt` target option which can be customized for each target. It is used to help determine a default for the PLT option (and `-Z plt` always overrides the setting).

bors commented:
:hourglass: Testing commit 6009da079419c9693fe4965ecacbd473c2553173 with merge 77af314083e5acabf9ba5335e47271f35eef2e99...

alexcrichton commented:
FWIW this looks like it may cause bugs in LLVM on i686-unknown-linux-gnu. I noticed that [stdsimd's CI](https://travis-ci.org/rust-lang-nursery/stdsimd) was [failing](https://travis-ci.org/rust-lang-nursery/stdsimd/jobs/443791815) for i686-unknown-linux-gnu because rustc was segfaulting. Some local investigation showed a segfault in LLVM. We compile that target with `-C relocation-model=static` on CI, and I believe the combination of 32-bit Linux with `-C relocation-model=static` was causing the issue. I haven't had a chance to dig deeper. I've worked around it with `-Z plt=yes`

eddyb commented:
FWIW, this broke pretty badly on certain (older) distro toolchains, but because `dylib` is largely unused, it took someone (ab)using `proc_macro::bridge` to run proc macros outside of `rustc` to trigger it: #61539.

MaskRay commented:
I think `-Z plt=no` is not a good default. Sent https://github.com/rust-lang/rust/pull/106380 to disable it. (Thanks to @GabrielMajeri for mentioning this review as I cannot find it with the commits).

If you see positive benchmark results, it is likely because dynamically linked libc calls dominate. If one statically links libc, `-Z plt=no` is going to be a pessimization.
Plus, x86-32 requires very new lld (as a maintainer, I just added support for `___tls_get_addr`; older lld will create a silently corrupted executable).
For GNU ld, a relatively new one is needed: 2016-06 https://sourceware.org/git/?p=binutils-gdb.git;a=commit;h=e2cbcd9156d1606a9f2153aecd93a89fe6e29180 (and a counterpart for x86-32)
